### PR TITLE
feat: encapsulated transactions API

### PR DESCRIPTION
**Pull request by adamaltman**
This is an idea for discussion.

### CODE REVIEW

File: openapi/components/schemas/PaymentInstruction.yaml [hunk 1]
```diff
@@ -2,3 +2,5 @@ oneOf:
   - $ref: PaymentInstructions/PaymentToken.yaml
   - $ref: PaymentInstructions/PaymentInstrument.yaml
   - $ref: PaymentInstructions/PaymentMethods.yaml
+  - $ref: PaymentCards/PaymentCardCreatePlain.yaml
```
Comment (colinbird) [inline openapi/components/schemas/PaymentInstruction.yaml:5]: `billingAddress` is required here, should it be used to create the customer if the root `billingAddress` is not provided?

Comment (nanobash) [reply openapi/components/schemas/PaymentInstruction.yaml:5]: Could it be that param is mandatory/optional based on other param bool value (conditional pretty much)?

`billingAddress` should be mandatory if customer does not exist, but if he exist, the field could be still optional

File: openapi/components/schemas/Transactions/TransactionRequest.yaml [hunk 2]
```diff
@@ -1,8 +1,12 @@
 allOf:
-  - $ref: ../Common/CommonTransactionRequest.yaml
   - required:
     - type
   - properties:
+      upsertCustomer:
+        type: boolean
+        default: false
+        writeOnly: true
+        description: If true it will upsert (create or update) a customer. If false the `customerId` must already exist.
```
[the remark this answers is not in the document]
Comment (nanobash) [reply openapi/components/schemas/Transactions/TransactionRequest.yaml:9]: Why? I think it's a good idea to generate object external id's (UUIDs) inside our system and not be dependant on provided params

[the remark this answers is not in the document]
Comment (VladimirBerdnik) [reply openapi/components/schemas/Transactions/TransactionRequest.yaml:9]: > Maybe allow to use customerId: null to make it auto-generated?

I've tried to say the same

File: openapi/components/schemas/Transactions/TransactionRequest.yaml [hunk 3]
```diff
@@ -20,7 +24,24 @@ allOf:
...
+        nullable: true
+        properties:
+          amount:
+            type: number
+            format: double
+            example: 275.35
+            description: The limit amount.
+          currency:
+            $ref: ../CurrencyCode.yaml
+          resetTime:
```
[the remark this answers is not in the document]
Comment (ashkarpetin) [reply openapi/components/schemas/Transactions/TransactionRequest.yaml:44]: > format: date-time

File: openapi/components/schemas/Transactions/TransactionRequest.yaml [hunk 4]
```diff
@@ -20,7 +24,24 @@ allOf:
...
+  - properties:
+      limits:
+        title: AmountLimits
+        type: object
+        nullable: true
+        properties:
+          amount:
+            type: number
+            format: double
+            example: 275.35
```
[the remark this answers is not in the document]
Comment (ashkarpetin) [reply openapi/components/schemas/Transactions/TransactionRequest.yaml:40]: No, this is correct

File: openapi/components/schemas/Transactions/TransactionRequest.yaml [hunk 5]
```diff
@@ -20,7 +26,24 @@ allOf:
           If the instruction is to `do-nothing`, a transaction with result `approved` of type `setup` will be returned.
         type: string
         enum:
-          - 3ds-authentication
           - sale
           - authorize
           - setup
+  - $ref: ../Common/CommonTransactionRequest.yaml
+  - properties:
```
Comment (ashkarpetin) [inline openapi/components/schemas/Transactions/TransactionRequest.yaml:33]: Please extract to separate schema for DRY

File: openapi/components/schemas/Transactions/Transaction.yaml [hunk 6]
```diff
@@ -253,6 +253,22 @@ allOf:
         readOnly: true
         nullable: true
         x-sortable: true
+      limits:
```
Comment (ashkarpetin) [inline openapi/components/schemas/Transactions/Transaction.yaml:256]: Please extract to separate schema for DRY.

Comment (nanobash) [reply openapi/components/schemas/Transactions/Transaction.yaml:256]: Extracted limits property to separate ResourceTransactionLimit.yaml file, thanks

File: openapi/components/schemas/ResourceTransactionLimit.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
+type: object
```
Comment (ashkarpetin) [inline openapi/components/schemas/ResourceTransactionLimit.yaml:1]: Plese remove Resource prefix.

File: openapi/components/schemas/Transactions/TransactionRequest.yaml [hunk 8]
```diff
@@ -1,8 +1,14 @@
...
   - required:
     - type
   - properties:
+      upsertCustomer:
+        type: boolean
+        default: false
+        writeOnly: true
+        description: If true it will upsert (create or update) a customer. If false the `customerId` must already exist.
+      customerId:
+        description: The customer identifier string. Optional if upserCustomer true.
```
Comment (adamaltman) [inline openapi/components/schemas/Transactions/TransactionRequest.yaml:11]:
```suggestion
        description: The customer identifier string. Optional if `upsertCustomer` is true.
```

**Review by adamaltman — COMMENTED** [review]
What is the reason for the customerId to not always be required?

**Comment by adamaltman** [comment]
This has been changed from my original proposal to the point where it has lost the spirit of it. As it is now, it won't meet the use case I was looking for. Let's discuss how to get it back on track.

**Review by nanobash — DISMISSED** [review]

**Comment by swilczynski** [comment]
@Rebilly/docs please re-review

File: openapi/components/schemas/Transactions/TransactionRequest.yaml [hunk 9]
```diff
@@ -24,10 +25,10 @@ allOf:
           If the instruction is to `do-nothing`, a transaction with result `approved` of type `setup` will be returned.
         type: string
         enum:
-          - 3ds-authentication
           - sale
           - authorize
           - setup
       limits:
         $ref: ../TransactionLimitAmount.yaml
   - $ref: ../Common/CommonTransactionRequest.yaml
```
Comment (ashkarpetin) [inline openapi/components/schemas/Transactions/TransactionRequest.yaml:33]: I see CommonTransactionRequest is included twice, that probably is a mistake

**Review by ashkarpetin — APPROVED** [review]

**Review by colinbird — APPROVED** [review]

**Pull request merged**